libcasr/src/stacktrace.rs

```diff
@@ -287,11 +288,226 @@ impl Filter for Stacktrace {
+            });
+        }
+        let mut keep = indices.iter();
+
+        // Remove repeating intervals and trusted functions from stack trace
         self.retain(|entry| {
-            (entry.function.is_empty() || !rfunction.is_match(&entry.function))
+            *keep.next().unwrap()
+                && (entry.function.is_empty() || !rfunction.is_match(&entry.function))
                 && (entry.module.is_empty() || !rfile.is_match(&entry.module))
                 && (entry.debug.file.is_empty() || !rfile.is_match(&entry.debug.file))
         });
     }
 }
+
+/// Map indices i of arr to the length of the longest border of arr[1 .. i]
+///
+/// # Arguments
+///
+/// * `arr` - given sequence
+///
+/// # Return value
+///
+/// Vector of mapped indices
+fn prefix_function<T: PartialEq>(arr: &[T]) -> Vec<usize> {
+    let mut z = Vec::new();
+    let (n, mut r, mut l) = (arr.len(), 0, 0);
+    z.resize(n, 0);
+    for i in 1..n {
+        if i <= r {
+            z[i] = z[i - l].min(r - i + 1);
+        }
+        while i + z[i] < n && arr[z[i]] == arr[i + z[i]] {
+            z[i].add_assign(1);
+        }
+        if i + z[i] - 1 > r {
+            l = i;
+            r = i + z[i] - 1;
+        }
+    }
+    z
+}
+
+/// Add interval to the resulting vector
+///
+/// # Arguments
+///
+/// * `intervals` - resulting vector
+///
+/// * `shift` - shift
+///
+/// And other algorithm parameters.
+/// More about Main-Lorentz algorithm https://cp-algorithms.com/string/main_lorentz.html
+fn add_interval(
+    intervals: &mut Vec<(usize, usize, usize)>,
+    shift: usize,
+    left: bool,
+    cntr: usize,
+    l: usize,
+    k1: usize,
+    _k2: usize,
+) {
+    let l1 = k1.min(l);
+    let l2 = l - l1;
+    let pos = if left {
+        cntr - l1
+    } else {
+        cntr + 1 - 2 * l1 - l2
+    };
+    let (left, right) = (shift + pos, shift + pos + 2 * l - 1);
+    let interval_len = right - left + 1;
+    if intervals
+        .iter()
+        .filter(|x| !(x.1 < left || right < x.0))
+        .all(|x| x.2 < interval_len)
+    {
+        intervals.retain(|x| x.1 < left || right < x.0);
+        intervals.push((left, right, right - left + 1));
+    }
+}
+
+/// Perform Main-Lorentz algorithm
+///
+/// # Arguments
+///
+/// * `s` - given sequence
+///
+/// * `shift` - shift
+///
+/// * `intervals` - the resulting vector consisting of tuples such as (interval start, interval end, period length)
+///
+/// * `nae` - element that not equal to others
+///
+/// More about Main-Lorentz algorithm https://cp-algorithms.com/string/main_lorentz.html
+fn main_lorentz<T: PartialEq>(
+    s: &[&T],
+    shift: usize,
+    intervals: &mut Vec<(usize, usize, usize)>,
+    nae: &T,
+) {
+    let n = s.len();
+    if n < 2 {
+        return;
+    }
+
+    let len_u = n / 2;
+    let len_v = n - len_u;
+    let (u, v): (Vec<&T>, Vec<&T>) = (s[..len_u].to_vec(), s[len_u..].to_vec());
+    let (mut ru, mut rv) = (u.clone(), v.clone());
```

Comment:
can we create a reverse iterator and call collect?

Reply:
done